### PR TITLE
Add Slack API caching layer

### LINKED ISSUE #105: Introduce caching
We might want to introduce caching of the values returned in `SlackAPI` by using TTL, for instance. And only for certain things, like users, channels etc., but not the responses to destructive events, like deleting files. One way to do this is adding an extra `ttl=None` argument to `SlackAPI.post` so each command has to control it directly. Another way is to control all of it inside `SlackAPI` and employ caching depending on the Slack API method used.

This module looks really useful: https://cachetools.readthedocs.io/en/latest/

Another thing to think about is whether we want the cache saved to disk or not? To keep things simple the cache would only be in memory.

### PR DESCRIPTION
**Pull request by voidpirate**
Fixes #105.

Changes proposed in this pull request:
- Add option to include local cache for each command using `TTLCache` from the `cachetools` library.
- Added methods to base command with defaults that can be overridden in command subclass to configure command local caching.
- Enabled caching for `users.list`.
- Debug logging for cache hit/miss

### CODE REVIEW

**Comment by voidpirate** [comment]
As a second step (or task rather) if the command has enabled caching a flag should be added dynamically that allows the user to clear the items in the cache.

File: slacker/slack_api.py [hunk 1]
```diff
@@ -148,3 +161,36 @@ def __validate_response(self, response):
...
+    """Get value from cache given the hash key"""
+    if self.__cache is None:
+      return None
+    return self.__cache.get(key)
+
+  def __update_cache(self, key, resp):
+    """Update cache value given a key"""
+    self.__logger.debug("Updating cache: {}".format(key))
+    self.__cache[key] = resp
+    return self.__cache[key]
```
Comment (netromdk) [inline slacker/slack_api.py]: The unnecessary lookup can be avoided by returning `resp` dierctly.

Comment (netromdk) [reply slacker/slack_api.py]: But maybe it shouldn't return anything?

File: slacker/slack_api.py [hunk 2]
```diff
@@ -148,3 +161,36 @@ def __validate_response(self, response):
...
+
+  def __generate_cache_key(self, url, params):
+    """Returns hash of the url and params"""
+    if self.__cache is None:
+      return None
+
+    m = hashlib.md5()
```
Comment (netromdk) [inline slacker/slack_api.py]: I'd prefer to use a SHA family hash instead, like SHA-256 - it doesn't have to be overkill.

File: slacker/slack_api.py [hunk 3]
```diff
@@ -66,9 +69,19 @@ def post(self, method, args={}):
     if self.__requires_token:
       args["token"] = self.__token
 
+    # Check for cached request
+    cache_key = self.__generate_cache_key(url, args)
+    if cache_key is not None and self.__is_cached(cache_key):
+      cache_value = self.__get_cached_value(cache_key)
+      if cache_value is not None:
+        return cache_value
```
Comment (netromdk) [inline slacker/slack_api.py]: Look-ups are made twice here. Instead, you could could do one look-up with `self.__get_cached_value` and use its return value as evidence of being cached or not. Afterwards, you can then simply use the value and no more look-ups are necessary.

**Comment by voidpirate** [comment]
> Looks great. The cache is living in memory only, right?

Yep, the cache is in-memory. If the command allows caching it will have its own `TTLCache`, cache will not be shared with other commands.

I've made the general improvements around how caching keys are generated and looked up.

Regarding clearing command caches. I agree with both of these approaches, perhaps we implement both? I think it is important that each command has the ability to clear its own cache when running with additional arguments. e.g. if you wanted to clear the cache for `users.list` so you make a fresh api call and cache the new response without executing two separate commands to clear the cache.

**Comment by netromdk** [comment]
This is great, thanks!

> Regarding clearing command caches. I agree with both of these approaches, perhaps we implement both?
Yeah, that's a good point. Should do that in another PR.

**Pull request merged**